### ISSUE TITLE
Cache-Writing Builds Produce Read-only Artifacts

### ISSUE THREAD

**Issue by TravisWhitaker** [ndmitchell/shake#701]
Consider this simple Shakefile.hs:
```Haskell
#!/usr/bin/env runhaskell

module Main where

import Control.Monad

import Development.Shake
import Development.Shake.FilePath
import Development.Shake.Util

main :: IO ()
main = shakeArgs (shakeOptions {shakeThreads = 0}) $ do

    want ["build/prog"]

    "build/prog" %> \o -> do
        ss <- getDirectoryFiles "" ["src/*.c"]
        let os = map (("build" </>) . (-<.> ".o")) ss
        need os
        unit $ command [] "clang" $ "-o" : o : os

    "build//*.o" %> \o -> do
        let c = dropDirectory1 (o -<.> ".c")
            m = o -<.> ".m"
        produces [m]
        unit $ command [] "clang" ["-c", c, "-o", o, "-MMD", "-MF", m]
        needMakefileDependencies m
```
With this small source tree:
```
.
├── Shakefile.hs
└── src
    ├── f.c
    └── main.c
```
After a build without the cache, the artifacts have the permissions one might expect:
```
.:
total 12
-rwxr-xr-x 1 travis staff 8456 Aug  2 21:52 prog
drwxr-xr-x 6 travis staff  192 Aug  2 21:52 src

./src:
total 16
-rw-r--r-- 1 travis staff  23 Aug  2 21:52 f.m
-rw-r--r-- 1 travis staff 620 Aug  2 21:52 f.o
-rw-r--r-- 1 travis staff  29 Aug  2 21:52 main.m
-rw-r--r-- 1 travis staff 812 Aug  2 21:52 main.o
```
However, after a build with `--share`, the artifacts are read-only:
```
.:
total 12
-r-xr-xr-x 2 travis staff 8456 Aug  2 21:54 prog
drwxr-xr-x 6 travis staff  192 Aug  2 21:54 src

./src:
total 16
-r--r--r-- 2 travis staff  23 Aug  2 21:54 f.m
-r--r--r-- 2 travis staff 620 Aug  2 21:54 f.o
-r--r--r-- 2 travis staff  29 Aug  2 21:54 main.m
-r--r--r-- 2 travis staff 812 Aug  2 21:54 main.o
```
This is rather annoying; if a source file changes, the build will fail since the compiler can't overwrite the artifact! Suppose main.c changes:
```
# clang (for build/src/main.o)
error: unable to open output file 'build/src/main.o': 'Operation not permitted'
1 error generated.
Error when running Shake build system:
  at want, called at Shakefile.hs:14:5 in main:Main
* Depends on: build/prog
* Depends on: build/src/main.o
* Raised the exception:
user error (Development.Shake.command, system command failed
Command line: clang -c src/main.c -o build/src/main.o -MMD -MF build/src/main.m
Exit code: 1
Stderr:
error: unable to open output file 'build/src/main.o': 'Operation not permitted'
1 error generated.
)
```
This is on macOS. Is this the expected behavior for `--share`?

**Comment by dwijnand** [comment]
> Is this the expected behavior for `--share`?

I think it is.  See https://github.com/ndmitchell/shake/commit/14e4385120d449abc3ba13bd8abcfbef0a53de1f:

> making files read only stops them from inadvertantly mutating the cache

Looks like Shake 0.18.3 (specifically https://github.com/ndmitchell/shake/commit/7aaddb51c359c857675ce349def3985bc3cd1311) introduced `--share-copy` to do direct `copyFile`ing, without any permission changing.

**Comment by TravisWhitaker** [comment]
Hmm, using symlinks is indeed the default, and it makes sense that symlinked artifacts should be read-only. This seems like a very strange default to me, as this makes the build tree essentially unusable for development, unless I'm somehow holding it wrong?

**Comment by dwijnand** [comment]
Perhaps it's organically grew like that?  Can `shakeSymlink` be switched to default `false`, @ndmitchell?  I assume that'll have to target Shake v0.19.

**Comment by TravisWhitaker** [comment]
Fixed in #718

**Comment by ndmitchell** [comment]
Thanks for spotting and closing! Note that I do not intend to make it Shake v0.19 since it's a new feature that I think most people who have used are probably suffering through, and thus its really a bug fix.
